llvm/lib/Transforms/Instrumentation/AddressSanitizer.cpp

```diff
@@ -1234,7 +1245,91 @@ static bool isUnsupportedAMDGPUAddrspace(Value *Addr) {
   return false;
 }
 
+static bool isUnsupportedSPIRAddrspace(Value *Addr, Function *Func) {
+  Type *PtrTy = cast<PointerType>(Addr->getType()->getScalarType());
+  if (PtrTy->getPointerAddressSpace() != 0) {
+    // Skip SPIR-V built-in varibles
+    StringRef Name;
+    if (auto *CE = dyn_cast<ConstantExpr>(Addr)) {
+      const auto OpCode = CE->getOpcode();
+      if (OpCode == Instruction::GetElementPtr ||
+          OpCode == Instruction::AddrSpaceCast) {
+        Name = CE->getOperand(0)->getName();
+      }
+    } else {
+      Name = Addr->getName();
+    }
+    if (Name.starts_with("__spirv_BuiltIn")) {
+      return true;
+    }
+  } else {
+    // FIXME: Check kernel arguments instead of value's name
+    auto Name = Addr->getName();
+    if (Name == "_arg_NumWorkItems" || Name == "_arg_KernelFunc") {
+      return true;
+    }
+  }
+  return false;
```

Comment:
I think that you can add several early exist into the function, reducing amount of indentation:

```c++
if (PtrTy->getPointerAddressSpace() == 0) { // private
  // FIXME: Check kernel arguments instead of value's name
  auto Name = Addr->getName();
  return Name == "_arg_NumWorkItems" || Name == "_arg_KernelFunc";
}

// All the rest address spaces: skip SPIR-V built-in varibles
StringRef Name;
if (auto *CE = dyn_cast<ConstantExpr>(Addr)) {
  const auto OpCode = CE->getOpcode();
  if (OpCode == Instruction::GetElementPtr ||
    OpCode == Instruction::AddrSpaceCast) {
    Name = CE->getOperand(0)->getName();
  }
} else
  Name = Addr->getName();

return Name.starts_with("__spirv_BuiltIn");
```

Swapping branches is up to you, it is just my personal preference to have shortest code branches first

Reply:
Thanks for you suggestion, it's much better than before